src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,566 @@
+            Integrator<Void, T, R> integrator,
+            BiConsumer<Void, Downstream<? super R>> finisher) {
+        return of(
+                defaultInitializer(),
+                integrator,
+                Gatherers.Value.DEFAULT.statelessCombiner,
+                finisher
+        );
+    }
+
+    /**
+     * Returns a new, parallelizable, {@code Gatherer} described by the given
+     * {@code initializer}, {@code integrator}, {@code combiner} and
+     * {@code finisher}.
+     *
+     * @param initializer the supplier function for the new gatherer
+     * @param integrator the integrator function for the new gatherer
+     * @param combiner the combiner function for the new gatherer
+     * @param finisher the finisher function for the new gatherer
+     * @param <T> the type of input elements for the new gatherer
+     * @param <A> the type of initializer for the new gatherer
+     * @param <R> the type of results for the new gatherer
+     * @throws NullPointerException if any argument is null
+     * @return the new {@code Gatherer}
+     */
+    static <T, A, R> Gatherer<T, A, R> of(
+            Supplier<A> initializer,
+            Integrator<A, T, R> integrator,
+            BinaryOperator<A> combiner,
+            BiConsumer<A, Downstream<? super R>> finisher) {
+        return new Gatherers.GathererImpl<>(
+                Objects.requireNonNull(initializer),
+                Objects.requireNonNull(integrator),
+                Objects.requireNonNull(combiner),
+                Objects.requireNonNull(finisher)
+        );
+    }
+
+    /**
+     * A Downstream object is the next stage in a pipeline of operations,
+     * to which elements can be sent.
+     * @param <T> the type of elements this downstream accepts
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Downstream<T> {
+
+        /**
+         * Pushes, if possible, the provided element to the next stage in the
+         * pipeline.
+         *
+         * <p>If this method returns {@code false} then the next stage does
+         * not accept any more elements.
+         *
+         * @param element the element to send
+         * @return {@code true} if more elements can be sent,
+         *         and {@code false} if not.
+         */
+        boolean push(T element);
+
+        /**
+         * Allows for checking whether the next stage is known to not want
+         * any more elements sent to it.
+         *
+         * @apiNote This is best-effort only, once this returns true it should
+         *          never return false again for the same instance.
+         *
+         * <p>By default this method returns {@code false}.
+         *
+         * @return {@code true} if this Downstream is known not to want any
+         *         more elements sent to it, {@code false} if otherwise
+         */
+        default boolean isRejecting() { return false; }
```

Comment:
Is it really a good default ?

Reply:
It is not better to have 2 Downstream (Downstream and a subtype) like you hae two integrators (Integrator and Greedy)

Reply:
Yes, it's the best possible default.

The Downstream can become rejecting after any push, so any operations which do multiple pushes would not see that.